Boehm GC for WebAssembly

aykevl opened the pull request:
See: https://github.com/tinygo-org/tinygo/pull/4385

This is an incomplete port. Things that need to be resolved are:

 - [x] #4385 needs to be merged first, and this PR needs to be rebased.
 - [x] We need to have two separate builds of wasi-libc since the `MALLOC_IMPL` macro is going to be different. I think it's easiest to start doing this build not when creating a TinyGo tarball but when running TinyGo itself (similar to picolibc, musl, etc). This makes it easier to build different variants easily and when needed.

The performance improvements can be large. For example, for the html package some benchmarks are more than 50% faster:

```
$ benchstat html-precise.txt html-boehm.txt
               │ html-precise.txt │           html-boehm.txt           │
               │      sec/op      │   sec/op     vs base               │
Escape               110.29µ ± 0%   73.62µ ± 3%  -33.25% (p=0.002 n=6)
EscapeNone            48.37µ ± 0%   49.15µ ± 1%   +1.62% (p=0.002 n=6)
Unescape             192.31µ ± 5%   84.53µ ± 3%  -56.04% (p=0.002 n=6)
UnescapeNone          2.638µ ± 1%   2.559µ ± 4%        ~ (p=0.132 n=6)
UnescapeSparse        45.44µ ± 2%   18.55µ ± 1%  -59.17% (p=0.002 n=6)
UnescapeDense        182.55µ ± 1%   79.63µ ± 1%  -56.38% (p=0.002 n=6)
geomean               53.11µ        32.40µ       -39.00%
```

Pause times should also be reduced quite a bit, likely much more than the performance improvements. I did not test those, though.

If you want to try this PR, you will have to rebuild wasi-libc:

```
$ rm -rf lib/wasi-libc/sysroot/ lib/wasi-libc/build
$ make wasi-libc
```

(Also note that only wasip1 is supported right now, other wasm variants should be easy to add in the future).

aykevl committed:
wasm: add Boehm GC support

This adds support for `-gc=boehm` on `-target=wasip1` and `-target=wasm`
(in a browser or NodeJS). Notably it does *not* add Boehm GC support for
`-target=wasip2`, since that target doesn't have a real libc.

deadprogram commented:
@aykevl you can now rebase this PR against `dev` now that #4385 was merged. Thanks!

aykevl commented:
A few more changes are needed to make this PR ready, it needs more than just rebasing. I'm working on it!

aykevl commented:
Should be ready now.
The Boehm GC is disabled by default, it can be used with `-gc=boehm`.

aykevl commented:
The main drawback right now is that it will often output messages like the following:

```
GC Warning: Repeated allocation of very large block (appr. size 68 KiB):
        May lead to memory leak and poor performance
```

I think the only reasonable fix is to just disable output entirely (see https://github.com/ivmai/bdwgc/pull/721) or disable it using `GC_set_warn_proc` since using `GC_malloc_ignore_off_page` is not safe with Go (inner pointers are allowed, even far into an object, so `GC_malloc_ignore_off_page` is unsafe even if it would probably work for most programs in practice).

deadprogram commented:
@aykevl now that https://github.com/ivmai/bdwgc/commit/6d10870ba8ca4410276af90c7fc305f1d2949c00 was merged, I think that this PR can proceed?

deadprogram commented:
@aykevl reminder about this PR :smiley_cat:

aykevl commented:
Updated the PR.
* rebased on dev
* added `GC_DISABLE_SNPRINTF=1` on WebAssembly to avoid linking in printf related functions (saves ~18kB of binary size)
* disabled warnings on WebAssembly since they happen frequently and there's not much we can do about them.

deadprogram commented:
@aykevl looks like actual error here: https://github.com/tinygo-org/tinygo/actions/runs/15212794972/job/42791925530?pr=4812#step:17:24

aykevl commented:
@deadprogram I think it's more likely to be a flaky test. I don't see how this PR should affect windows/386.

aykevl commented:
@deadprogram actually, you might be right. The Boehm GC is also used on Windows, and this PR updates the bdwgc library.

aykevl commented:
I can reproduce this issue in my Windows VM (running on ARM!).

aykevl commented:
I think it's related to https://github.com/ivmai/bdwgc/issues/736. Probably our bug and not one in bdwgc, but asking to be sure.

EDIT: might not be the actual GOOS=386 issue here.

aykevl commented:
Oh it is actually https://github.com/ivmai/bdwgc/commit/e8aedf635e1ad42102799478ea8a2ab55abd2ba5 that is causing this issue on Windows.

aykevl commented:
Pushed a workaround to disable bdwgc on Windows/386 for now. I don't think that's a big loss, I think having bdwgc on WASIp1 is much more important.

deadprogram commented:
@aykevl not quite passing yet https://github.com/tinygo-org/tinygo/actions/runs/15388465010/job/43292271205#step:17:24

aykevl commented:
Should be fixed now.

deadprogram commented:
OK, new error https://github.com/tinygo-org/tinygo/actions/runs/15389453659/job/43296579801?pr=4812#step:7:28

aykevl commented:
I can't reproduce this issue locally. Trying to re-run the tests to see whether it's a flaky one.

deadprogram commented:
> I can't reproduce this issue locally. Trying to re-run the tests to see whether it's a flaky one.

That time it failed on yet a different test. I restarted one more time...

deadprogram commented:
It fails on a different test each time, looking like some kind of race? In any case, still fails eventually on some test or another. :crying_cat_face:

deadprogram commented:
Passing all tests in the last run. Whatever race condition is happening, I do not think it is related to this PR.

deadprogram reviewed (approved):
From my point of view, this PR appears ready to merge now.

Anyone have any further feedback before I do so?

aykevl commented:
I'm not so sure about that. I think you just managed to avoid a somewhat rare crash this time, but the bug is still there.

deadprogram commented:
> I'm not so sure about that. I think you just managed to avoid a somewhat rare crash this time, but the bug is still there.

@aykevl I don't think that is specific to this PR, do you?

aykevl commented:
> I don't think that is specific to this PR, do you?

I think it is, since we were also hitting issues on windows/386. I think the windows/386 issues and the flaky builds are related.

Currently trying to reproduce it on my system.

aykevl commented:
Changed to use the bdwgc commit just before https://github.com/ivmai/bdwgc/commit/e8aedf635e1ad42102799478ea8a2ab55abd2ba5 which I think introduced the Windows issues (GOARCH=386 at least, but I also suspect it is causing the flaky CI issue). See https://github.com/ivmai/bdwgc/issues/736#issuecomment-2929578337 for details. This does mean we lose some of the recent size improvements, but fixing these issues will need a bit more time. To at least have (opt-in) bdwgc support in wasm for this release (even if it grows the binary more than needed) is better than having nothing.

Also, this brings back bdwgc support in GOARCH=386, which isn't super useful but a nice side effect I guess.

deadprogram commented:
> To at least have (opt-in) bdwgc support in wasm for this release (even if it grows the binary more than needed) is better than having nothing.

Agreed!

aykevl commented:
And all tests pass in the first CI run, this gives me some confidence I was right about that one bdwgc commit being a problem.

deadprogram commented:
Thanks for working on this @aykevl and the various needed fixes along the way. Now merging.
